### PR TITLE
Upgrade previous "Extensions" system to new "Errors" model

### LINKED ISSUE #18796: Don't check errors against instances
There's a couple places where we're checking `if (x instanceof BaseException)`. When a different version of the library where `BaseException` comes from is used, this check fails. Now that we're starting to create errors through a new abstraction, we should change that to an explicit check against the error code instead: `if (isDirectusError(x) && x.code === 'CODE')`

### PR DESCRIPTION
**Pull request by rijkvanzanten**
This PR introduces a new `@directus/errors` package that can be used to easily create your own "Directus-flavored" errors from within extensions, to make sure you can throw errors that Directus' error handler understands and can handle properly. This is a replacement for the previously un-typed exceptions that were shared with Hooks.

This PR also updates the old `BaseException`-based errors with new Errors that are based on this same new system.

Fixes https://github.com/directus/directus/issues/18796

## Other notable changes

- Also creates a new (mostly empty) `@directus/validation` package we can start moving the validation specific logic into[^1]
- Deprecates `@directus/exceptions`
- Exceptions are no longer passed to Extensions, instead, they should create bespoke errors through `@directus/errors`

## Todo

I still have a couple more errors to move over, and some build issues to resolve, but I'd love an initial review / questions early 🚀 

- [x] Finalize last couple errors
- [x] Fix build
- [x] Fix tests
- [x] Update extensions documentation
- [x] Add changeset

[^1]: There was a "convert-joi-to-directus-error" type function hacked into a shared exception that's now cleaned up

### CODE REVIEW

**Review by licitdev — COMMENTED** [review]
Should the error codes and service names be extracted to `@directus/constants` to improve consistency?

Eg: The `FORBIDDEN` error code
https://github.com/directus/directus/blob/6ecdf13f786d59d270f79827b473d34d99f99be2/api/src/errors/forbidden.ts#L3

https://github.com/directus/directus/blob/6ecdf13f786d59d270f79827b473d34d99f99be2/api/src/controllers/activity.ts#L108

Eg: The `api` service name
https://github.com/directus/directus/blob/6ecdf13f786d59d270f79827b473d34d99f99be2/api/src/app.ts#L125

**Comment by rijkvanzanten** [comment]
@licitdev 

> Should the error codes and service names be extracted to @directus/constants to improve consistency?
> 
> Eg: The FORBIDDEN error code

I don't think so. We should aim to keep these packages scoped by functionality, rather than by "type". Eventually, I can imagine that the error types for websockets will live in a constant in `@directus/ws`, but not in `@directus/constants` 🤔

File: packages/validation/src/errors/failed-validation.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			message += ` Value has to be empty.`;
+			break;
+		case 'nempty':
+			message += ` Value can't be empty.`;
+			break;
+		case 'required':
+			message += ` Value is required be empty.`;
+			break;
+		case 'regex':
+			message += ` Value doesn't have the correct format.`;
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply packages/validation/src/errors/failed-validation.ts:85]: This is currently just matching the language we already use in the app. We can optimize in a separate pr 🙂

**Comment by licitdev** [comment]
> the error types for websockets will live in a constant in `@directus/ws`, but not in `@directus/constants`

Yes that would be possible. 👍🏼 Hmm, or could it be a constant within the errors package?

**Comment by rijkvanzanten** [comment]
@licitdev I wanna make sure we only put the stuff in those "global" packages that's truly "global". F.e. we should prevent adding a "GraphqlValidationError" in `@directus/errors`, but we could consider moving the generic "ForbiddenError" in there (as that's shared / reused in multiple contexts) 🤔

**Review by azrikahar — COMMENTED** [review]
Additional changes done during the review:

- replaced existing hardcoded `error.code === 'FORBIDDEN'` to `error.code === ErrorCode.Forbidden` in https://github.com/directus/directus/pull/18797/commits/e145268ee7ba0ff59c00445a230564412eebaea5
- fixed blackbox auth `login.test.ts` failing due to expecting `GRAPHQL_VALIDATION_EXCEPTION` instead of the newer `GRAPHQL_VALIDATION`, which was updated in https://github.com/directus/directus/commit/6ecdf13f786d59d270f79827b473d34d99f99be2

File: api/src/websocket/authenticate.ts [hunk 2]
```diff
@@ -39,11 +38,7 @@ export async function authenticateConnection(
 		const expires_at = getExpiresAtForToken(access_token);
 		return { accountability, expires_at, refresh_token } as AuthenticationState;
 	} catch (error) {
-		if (error instanceof InvalidCredentialsException && error.message === 'Token expired.') {
-			throw new WebSocketException('auth', 'TOKEN_EXPIRED', 'Token expired.', message['uid']);
-		}
```
[the remark this answers is not in the document]
Comment (br41nslug) [reply api/src/websocket/authenticate.ts:44]: Yep checked that and the new TokenExpired error gets properly caught and thrown as websocket error further down the line 😄

File: packages/errors/src/create-error.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	message: string | ((extensions: Extensions) => string),
+	status = 500
+): DirectusErrorConstructor<Extensions> => {
+	return class extends Error implements DirectusError<Extensions> {
+		override name = 'DirectusError';
+		extensions: Extensions;
+		code = code.toUpperCase();
+		status = status;
+
+		constructor(extensions: Extensions, options?: ErrorOptions) {
```
Comment (jaads) [inline packages/errors/src/create-error.ts:23]: `ErrorOptions` seems to be not defined 🤔

**Comment by azrikahar** [comment]
I believe we'll also need to add a copy of MIT `license` file to each of `@directus/errors` and `@directus/validation`.

File: api/src/errors/invalid-config.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { createError } from '@directus/errors';
+import { ErrorCode } from './codes.js';
+
+export interface InvalidConfigErrorExtensions {
+	provider: string;
+	reason?: string;
+}
+
+export const InvalidConfigError = createError<InvalidConfigErrorExtensions>(
```
Comment (paescuj) [inline api/src/errors/invalid-config.ts:9]: Should we rename this Error to something like `InvalidProviderError` as it seems very specific to Provider and not Config?

File: packages/errors/package.json [hunk 5]
```diff
@@ -21,18 +25,18 @@
...
+		"@types/ms": "0.7.31",
+		"@types/node": "18.16.12",
 		"@vitest/coverage-c8": "0.31.1",
-		"joi": "17.9.2",
+		"tsup": "6.7.0",
 		"typescript": "5.0.4",
 		"vitest": "0.31.1"
+	},
+	"dependencies": {
+		"ms": "2.1.3"
```
Comment (paescuj) [inline packages/errors/package.json:40]: A remnant?

File: api/src/controllers/activity.ts [hunk 6]
```diff
@@ -103,7 +104,7 @@ router.post(
 				data: record || null,
 			};
 		} catch (error: any) {
-			if (error instanceof ForbiddenException) {
+			if (isDirectusError(error) && error.code === ErrorCode.Forbidden) {
```
Comment (licitdev) [inline api/src/controllers/activity.ts:107]: There are frequent error code equality checks alongside `isDirectusError()`.

```ts
isDirectusError(error) && error.code === ErrorCode.Forbidden
```

Should the `isDirectusError()` be refactored to include an optional parameter for the error code equality check?

```ts
isDirectusError(error, ErrorCode.Forbidden)
```

Comment (rijkvanzanten) [reply api/src/controllers/activity.ts:107]: Love that idea! Good call

File: packages/errors/src/create-error.ts (hunk 3 above)
Comment (paescuj) [reply packages/errors/src/create-error.ts:23]: Type comes from default lib, so it should be available 👍 Maybe dependencies weren't installed on your end? 🤔

**Review by paescuj — APPROVED** [review]
error

**Pull request merged**